### PR TITLE
Add simple pickle compatibility to ReferenceFrame

### LINKED ISSUE #279: Problem dumping frames in sympy.physics.mechanics
Hi, 
I use sympy to find equations of motion . I can dump and load any sympy expressions except those containing frames . A simple example would be  : 
 
```
import sympy as sm
import sympy.physics.mechanics as me
import dill 

frame = me.ReferenceFrame('frame')
 
with open('test.pkl', 'wb') as file :
    dill.dump(frame, file)

with open('test.pkl', 'rb') as file :
    frame_=dill.load(file)

    assert frame_== frame
```

dumping works just fine but I get problem loading it  : 
new__() takes exactly 4 arguments (2 given)

I was wondering if there is a workaround for this problem ?

capture

### PR DESCRIPTION
**Pull request by tjstienstra**
#### References to other Issues or PRs

Partially fixes [#279 from dill](https://github.com/uqfoundation/dill/issues/279).

#### Brief description of what is fixed or changed
Implement a `__getnewargs_ex__` in `CoordinateSym` to enable some simple pickling of reference frames. One should however note that a `ReferenceFrame` is not exactly the same after de-serialization as before, due to the hash being based on the memory location. Another problem is that `dynamicsymbols` are not pickable, as described in #4297 (this last problem seems to be fixed in `cloudpickle`). An option I looked into for solving the hash problem, namely creating a `_hash` attribute did not work well, due to having a bi-directional graph that uses dictionaries (and so the hashes).

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/physics/vector/tests/test_frame.py [hunk 1]
```diff
@@ -720,3 +721,13 @@ def test_unit_dyadic():
...
+
+
+def test_pickle_frame():
+    N = ReferenceFrame('N')
+    A = ReferenceFrame('A')
+    A.orient_axis(N, N.x, 1)
+    A_C_N = A.dcm(N)
+    N1 = pickle.loads(pickle.dumps(N))
+    A1 = tuple(N1._dcm_dict.keys())[0]
+    assert A1.dcm(N1) == A_C_N
```
Comment (moorepants) [inline sympy/physics/vector/tests/test_frame.py:733]: The assertion only checks that these two matrices are the same but you pickle the reference frame. Shouldn't you be checking reference frames before and after?

Comment (moorepants) [reply sympy/physics/vector/tests/test_frame.py:733]: ok, i don't think this hurts anything, so let's merge

**Comment by TJStienstra** [comment]
Thanks for merging

File: sympy/physics/vector/tests/test_frame.py (hunk 1 above)
Comment (moorepants) [reply sympy/physics/vector/tests/test_frame.py:733]: At the minimum you could open an issue or add to one of the existing issues an explanation of the further complexity.

**Pull request merged**